examples/rust/tcp-push-pop.rs

```diff
@@ -144,24 +120,50 @@ impl TcpServer {
             };
 
             // Sanity check received data.
-            let ptr: *mut u8 = self.sga.expect("should be a valid sgarray").sga_segs[0].sgaseg_buf as *mut u8;
-            let len: usize = self.sga.expect("should be a valid sgarray").sga_segs[0].sgaseg_len as usize;
-            let slice: &mut [u8] = unsafe { slice::from_raw_parts_mut(ptr, len) };
+            let ptr: *mut u8 = sga.sga_segs[0].sgaseg_buf as *mut u8;
+            let bytes: usize = sga.sga_segs[0].sgaseg_len as usize;
+            debug_assert_eq!(bytes, BUF_SIZE_BYTES);
+            debug_assert!(ptr.is_aligned());
+            debug_assert_eq!(ptr.is_null(), false);
+
+            let slice: &mut [u8] = unsafe { slice::from_raw_parts_mut(ptr, BUF_SIZE_BYTES) };
 
             for x in slice {
-                demikernel::ensure_eq!(*x, fill_char);
+                demikernel::ensure_eq!(*x, i as u8);
             }
 
-            i += self.sga.expect("should be a valid sgarray").sga_segs[0].sgaseg_len as usize;
-
-            match self.libos.sgafree(self.sga.expect("should be a valid sgarray")) {
-                Ok(_) => self.sga = None,
-                Err(e) => anyhow::bail!("failed to release scatter-gather array: {:?}", e),
-            }
-            println!("pop {:?}", i);
+            self.libos.sgafree(sga)?;
+            println!("pop {:?} bytes", i * BUF_SIZE_BYTES);
         }
+        Ok(())
+    }
+
+    pub fn run(&mut self, local_socket_addr: SocketAddr) -> Result<()> {
+        if let Err(e) = self.libos.bind(self.listening_sockqd, local_socket_addr) {
+            anyhow::bail!("bind failed: {:?}", e.cause)
+        };
+
+        if let Err(e) = self.libos.listen(self.listening_sockqd, 16) {
+            anyhow::bail!("listen failed: {:?}", e.cause)
+        };
+
+        let qt: QToken = match self.libos.accept(self.listening_sockqd) {
+            Ok(qt) => qt,
+            Err(e) => anyhow::bail!("accept failed: {:?}", e.cause),
+        };
+
+        self.accepted_sockqd = match self.libos.wait(qt, Some(TIMEOUT_SECONDS)) {
+            Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_ACCEPT => unsafe { Some(qr.qr_value.ares.qd.into()) },
+            Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_FAILED => anyhow::bail!("accept failed: {}", qr.qr_ret),
+            Ok(qr) => anyhow::bail!("unexpected opcode: {:?}", qr.qr_opcode),
+            Err(e) => anyhow::bail!("operation failed: {:?}", e.cause),
+        };
+
+        // Wait for blocking pops.
+        self.pop_rounds()?;
+        // Wait for non-blocking pops.
```

Comment:
Is this ordering implicit - how is it ensured? i.e., the first time `pop_rounds` is called, it will wait for all blocking pops and next time for all the non-blocking pops?

Reply:
The order is imposed by the TCP connection because we're pushing all of the blocking requests and then the non-blocking ones.